### PR TITLE
[SYCL][CUDA] Fix LIT testing with CUDA devices

### PR DESCRIPTION
**Pull request by bjoernknafla**
Build the `get_device_count_by_type` tool with the required preprocessor
definitions to enable CUDA, and link with CUDA.

Passed correct backend name to the `get_device_count_by_type` tool to
is required to run tests with PI OpenCL instead of PI CUDA.

Clean up the `get_device_count_by_type` tool code to make it harder to
call it with the wrong backend and get unexpected query results back.

Signed-off-by: Bjoern Knafla <bjoern@codeplay.com>

### CODE REVIEW

**Comment by bjoernknafla** [comment]
**WARNING** This PR fixes CUDA LIT testing and multiple LIT tests will unexpectedly pass while more will fail. A follow up PR will fix this as I didn't want to overload this PR.

File: sycl/test/lit.cfg.py [hunk 1]
```diff
@@ -150,6 +150,8 @@ def getDeviceCount(device_type):
     config.available_features.add('gpu')
     if cuda:
        config.available_features.add('cuda')
+       gpu_run_substitute += " SYCL_BE=PI_CUDA "
```
Comment (bjoernknafla) [inline sycl/test/lit.cfg.py]: Adapt/rebase on top of https://github.com/intel/llvm/pull/1288 instead which has a more elegant and simpler solution.

File: sycl/test/lit.cfg.py [hunk 2]
```diff
@@ -106,13 +106,13 @@ def getDeviceCount(device_type):
         if len(result) > 1 and len(result[1]):
             print("getDeviceCount {TYPE}:{MSG}".format(
                 TYPE=device_type, MSG=result[1]))
-            if re.match(r".*cuda", result[1]):
+            if re.match(r".*cuda.*", result[1]):
```
Comment (bjoernknafla) [inline sycl/test/lit.cfg.py]: I am not sure if this is needed or not.

File: sycl/tools/get_device_count_by_type.cpp [hunk 3]
```diff
@@ -9,106 +9,204 @@
...
+              << help << std::endl;
+    return EXIT_FAILURE;
+  }
+
+  std::string msg;
+  cl_uint deviceCount = 0;
+
+  bool querySuccess = false;
+
+  if (backend == "opencl" || backend == "OpenCL" || backend == "OPENCL" ||
```
Comment (romanovvlad) [inline sycl/tools/get_device_count_by_type.cpp]: Suggest converting input string to lower case and check "opencl" substring existence.

File: sycl/tools/get_device_count_by_type.cpp [hunk 4]
```diff
@@ -9,106 +9,204 @@
...
+      backend == "PI_OPENCL") {
+    querySuccess = queryOpenCL(deviceType, deviceCount, msg);
+  } else if (backend == "cuda" || backend == "CUDA" || backend == "PI_CUDA") {
+    querySuccess = queryCUDA(deviceType, deviceCount, msg);
+  } else {
+    std::stringstream stream;
```
Comment (romanovvlad) [inline sycl/tools/get_device_count_by_type.cpp]: Why stream is required here? Can you just append string?

File: sycl/plugins/cuda/CMakeLists.txt [hunk 5]
```diff
@@ -6,7 +6,8 @@ message(STATUS "Including the PI API CUDA backend.")
 
 find_package(CUDA 10.0 REQUIRED)
 
-add_library(cudadrv SHARED IMPORTED)
+# Make imported library global to use it within the project.
+add_library(cudadrv SHARED IMPORTED GLOBAL)
```
Comment (romanovvlad) [inline sycl/plugins/cuda/CMakeLists.txt]: cudadrv is found by find_package(CUDA...) ?

File: sycl/tools/get_device_count_by_type.cpp [hunk 6]
```diff
@@ -9,106 +9,204 @@
 #include <CL/cl.h>
 #include <CL/cl_ext.h>
 
+#include <cstdlib>
```
Comment (romanovvlad) [inline sycl/tools/get_device_count_by_type.cpp]: Please, move this include to other standard C++ includes.

File: sycl/tools/get_device_count_by_type.cpp (hunk 3 above)
Comment (bjoernknafla) [reply sycl/tools/get_device_count_by_type.cpp]: Good point, will do.

**Comment by romanovvlad** [comment]
@bjoernknafla ping

File: sycl/test/lit.cfg.py (hunk 1 above)
Comment (bjoernknafla) [reply sycl/test/lit.cfg.py]: I decided to not adapt the mentioned [PR](https://github.com/intel/llvm/pull/1288) as setting the environment is not visible in LIT's output when a command fails and therefore makes it harder to copy and paste command invocation to then run and inspect by hand.

File: sycl/tools/get_device_count_by_type.cpp (hunk 6 above)
Comment (bjoernknafla) [reply sycl/tools/get_device_count_by_type.cpp]: Will do.

File: sycl/plugins/cuda/CMakeLists.txt (hunk 5 above)
Comment (bjoernknafla) [reply sycl/plugins/cuda/CMakeLists.txt]: The [FindCUDA package](https://cmake.org/cmake/help/v3.17/module/FindCUDA.html) (deprecated since CMake 3.10) does not find or export the CUDA driver.

The new [FindCUDAToolkit](https://cmake.org/cmake/help/v3.17/module/FindCUDAToolkit.html) (avialable from CMake 3.17 on) export the CUDA driver library.

File: sycl/tools/get_device_count_by_type.cpp (hunk 4 above)
Comment (bjoernknafla) [reply sycl/tools/get_device_count_by_type.cpp]: Good point, will simplify.

File: sycl/test/lit.cfg.py (hunk 2 above)
Comment (bjoernknafla) [reply sycl/test/lit.cfg.py]: It isn't. Removed the change.

**Comment by bjoernknafla** [comment]
I have updated the list of failing and now also hanging tests. I will first rework the PR that fixes LIT tests before this PR can go in.

**Comment by bader** [comment]
@bjoernknafla, could you take a look at merge conflicts, please?

**Comment by bjoernknafla** [comment]
@bader Will do.

File: sycl/test/lit.cfg.py [hunk 7]
```diff
@@ -119,11 +120,11 @@ def getDeviceCount(device_type):
 if getDeviceCount("cpu")[0]:
     found_at_least_one_device = True
     lit_config.note("Found available CPU device")
-    cpu_run_substitute = "env SYCL_DEVICE_TYPE=CPU "
+    cpu_run_substitute = "env SYCL_DEVICE_TYPE=CPU SYCL_BE={SYCL_BE} ".format(SYCL_BE=backend)
```
Comment (bjoernknafla) [inline sycl/test/lit.cfg.py:123]: On a fail LIT prints out this information and eases running the test by hand. Though the user still needs to add the `LD_LIBRARY_PATH` to find the SYCL library, but that is not part of this change.

File: sycl/tools/CMakeLists.txt [hunk 8]
```diff
@@ -5,45 +5,29 @@ set(CMAKE_CXX_EXTENSIONS OFF)
 add_executable(get_device_count_by_type get_device_count_by_type.cpp)
 add_dependencies(get_device_count_by_type ocl-headers ocl-icd)
 
-if( SYCL_BUILD_PI_CUDA )
```
Comment (bjoernknafla) [inline sycl/tools/CMakeLists.txt:8]: I removed this to avoid duplication and instead use the `cudadrv` target set up in the `sycl/plugin/cuda/CMakeLists.txt` CMake file.

File: sycl/tools/get_device_count_by_type.cpp [hunk 9]
```diff
@@ -11,105 +12,215 @@
...
+#include <sstream>
 #include <string>
 #include <vector>
 
 static const std::string help =
     "   Help\n"
     "   Example: ./get_device_count_by_type cpu opencl\n"
-    "   Supported device types: cpu/gpu/accelerator/default/all\n"
-    "   Supported backends: PI_CUDA/PI_OPENCL \n"
+    "   Support types: cpu/gpu/accelerator/default/all\n"
```
Comment (bjoernknafla) [inline sycl/tools/get_device_count_by_type.cpp:28]: Missed this - will fix.

**Comment by bader** [comment]
@romanovvlad, please, approve to unblock merge.

**Comment by bjoernknafla** [comment]
There is an open change request that GitHub cannot identify as the commit (due to rebasing) is gone...

**Comment by bader** [comment]
@bjoernknafla, basic_tests/buffer/buffer_dev_to_dev.cpp seems to hang. Do you know why?

**Comment by bjoernknafla** [comment]
I haven’t seen `buffer_dev_to_dev.cpp` hanging when running it locally (otherwise I wouldn’t have updated the PR). I will take a look (though not today).

**Comment by bader** [comment]
I've restarted the job and it passed.
I'll keep an eye on this check and if this issue repeats, I'll investigate.
